node/workflow-api.md

```diff
@@ -0,0 +1,322 @@
+# 3rd revision of the Typescript SDK Workflow API
+
+### Current API
+
+```ts
+import { Trigger } from '@temporalio/workflow';
+
+export type MyWorkflow = (
+  arg1: number,
+  arg2: string
+) => {
+  execute(): Promise<void>;
+  signals: {
+    unblock(): void;
+  };
+  queries: {
+    isBlocked(): boolean;
+  };
+};
+
+export const myWorkflow: MyWorkflow = (arg1, arg2) => {
+  const unblocked = new Trigger<void>();
+  let isBlocked = true;
+  return {
+    // Main workflow logic
+    async execute() {
+      await unblocked;
+      isBlocked = false;
+    },
+    signals: {
+      unblock() {
+        unblocked.resolve();
+      },
+    },
+    queries: {
+      isBlocked() {
+        return isBlocked;
+      },
+    },
+  };
+};
+```
+
+#### Issues
+
+- Types must be explicitly declared (To make [eslint happy](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/explicit-module-boundary-types.md))
+- Very verbose
+- Trigger (ResolvablePromise) is a foreign concept to JS developers
+- Trigger doesn't provide all of the functionality for this workflow, an additional `isBlocked` boolean variable is required
+- Signals and queries must be defined at the top level in order for types to be inferred.
+
+#### Advantages
+
+- Types for all Workflow handlers can be inferred at Workflow handle creation
+
+### Goals for the redesign
+
+- Maintain type inference capabilities - Mandatory for the Typescript SDK
+- Dynamic signals and queries registration first - A single API for Workflow handler registration
+- Componentization of related functionality - increase reusability
+- Implicit interface declarations - less boilerplate
+- No global variables - to support running Workflows in less isolated runtimes
+
+### New proposal
+
+#### Remove the `Trigger` API in favor of `condition`
+
+`condition` is similar to `Workflow.await` in Java, it returns a Promise and accepts a closure which resolve the Promise when it returns `true`.
+
+```ts
+import { condition } from '@temporalio/workflow';
+
+let isBlocked = true;
+await condition(() => !isBlocked);
+```
+
+#### Remove the Workflow factory wrapper and always use dynamic handler registration
+
+`workflows.ts`
+
+```ts
+import { Signal, Query, condition } from '@temporalio/workflow';
+
+// Define the signal / query with the type signature at the top level to share the definition for handle creation
+export const unblockSignal = new Signal<[] /* args */>('unblock');
+export const isBlockedQuery = new Query<[] /* args */, boolean /* return type */>('isBlocked');
+
+// Types are inferred without an additional type definition
+export async function myWorkflow(arg1: number, arg2: string): Promise<void> {
+  let isBlocked = true;
+  unblockSignal.handler = () => (isBlocked = false);
```

Comment:
issue raised by @vkarpov15 - what about awaiting for signal?

alternative proposal:

```
await unblockSignal // for oneoff signal
unblockSignal(() => {}) // for handler code/repeated signals
```

Reply:
And if unblockSignal can't be promisey itself, could do:
```js
import { Signal } from '@temporalio/workflow';

export const unblockSignal = new Signal<[]>('unblock');

export async function myWorkflow(arg1: number, arg2: string): Promise<void> {
  await unblockSignal.asPromise()
}
```

Reply:
I would heavily favor something like what @lorensr wrote. The signal itself being a promise you can await on is I guess neat, but kinda nonobvious. If you say "here's a signal" when you declare it, and "here's what it does with it fires" with `.handler`, then you probably want a "here's a wait to await on it" as an explicit concept as well. Maybe just `.fired()`

Reply:
If the issue we are addressing is that a Signal represents an awaitable with multiple firing, are we just replicating a generator?  I.e. is `yield` an appropriate semantic here?

Alternatively, if we rename `handler()` to `then()`, it suddenly seems a lot more like the normal Promise API, where both of the following are part of the standard ES6 API:
```
await myPromise;
myPromise.then(() => {});
```

Reply:
I think making signal awaitable or an async iterator is a separate discussion.
The basic API should be callback based and is flexible enough to support any use case.
To sum this up, these are the issues with use iterators and promises:
1. With an awaitable, how does waiting on multiple signals look like?
  - Does the promise always resolve on the first signal?
  - Is a new promise created when awaited?
  - Does the runtime buffer incoming signals until awaited?
2. With an async iterable, how does a user drain all signals?